### PR TITLE
BUG: df.getitem returning copy instead of view for unique column in dup index

### LINKED ISSUE #41062: BUG: df[frozenset] fails if (other) column names are not unique
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python
import pandas as pd

k = frozenset(["KEY"])

df1 = pd.DataFrame([[1,2,3,4], [5,6,7,8]], columns=[k, "B", "C", "D"])
df2 = pd.DataFrame([[1,2,3,4], [5,6,7,8]], columns=[k, "B", "C", "C"])

print(df1[k])  # works
print(df2[k])  # KeyError: "None of [Index(['KEY'], dtype='object')] are in the [columns]"

```

#### Problem description

Indexing a dataframe with a frozenset raises a KeyError if column names are not unique (the name of the indexed column is unique): 

```python-traceback
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\test\venv124\lib\site-packages\pandas\core\frame.py", line 3030, in __getitem__
    indexer = self.loc._get_listlike_indexer(key, axis=1, raise_missing=True)[1]
  File "C:\test\venv124\lib\site-packages\pandas\core\indexing.py", line 1266, in _get_listlike_indexer
    self._validate_read_indexer(keyarr, indexer, axis, raise_missing=raise_missing)
  File "C:\test\venv124\lib\site-packages\pandas\core\indexing.py", line 1308, in _validate_read_indexer
    raise KeyError(f"None of [{key}] are in the [{axis_name}]")
KeyError: "None of [Index(['KEY'], dtype='object')] are in the [columns]"
```

Indexing another dataframe that only differs by having unique column names works, so using a frozenset as column name seems to be ok. 

The traceback seems to indicate that a "listlike_indexer" is involved - maybe with non-unique column names a different indexer is used that has problems with non-atomic column names, turning the column name ``frozenset(["KEY"])`` into ``"KEY"``. Just a guess based on the error message, I don't know anything about how pandas indexers work.

#### Expected Output

Indexing works with a column name of arbitrary type regardless of whether the other column names are unique or not.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 2cb96529396d93b46abab7bbc73a208e708c642e
python           : 3.9.4.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.18362
machine          : AMD64
...

pandas           : 1.2.4
numpy            : 1.20.2
pytz             : 2021.1
dateutil         : 2.8.1
...

</details>

### LINKED ISSUE #45316: BUG: Series.replace not working if there are duplicated columns
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the master branch of pandas.

- [X] I have confirmed this bug exists on 1.4.0rc0.

### Reproducible Example

```python
import pandas as pd

df = pd.DataFrame([['a', 'x', 'o'], ['b', 'y', 'p'], ['c', 'z', 'q']], columns=['F', 'F', 'G'])
df['G'].replace('p', 'NEW', inplace=True)
print(df)
```

### Issue Description

When the DataFrame has duplicated columns names, the `Series.replace` with `inplace=True` of a not-duplicated column does not work (it does not perform the replacement) and raises a warning:
```
/home/arnau/.virtualenvs/lr/lib/python3.10/site-packages/pandas/core/generic.py:6619: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)
```

Some examples:
```
# doesn't work
df = pd.DataFrame([['a', 'x', 'o'], ['b', 'y', 'p'], ['c', 'z', 'q']], columns=['F', 'F', 'G'])
df['G'].replace('p', 'NEW', inplace=True)
```

```
# do work (because not inplace)
df = pd.DataFrame([['a', 'x', 'o'], ['b', 'y', 'p'], ['c', 'z', 'q']], columns=['F', 'F', 'G'])
df['G'] = df['G'].replace('p', 'NEW')
```

```
# do work (because there is no duplicated column)
df = pd.DataFrame([['a', 'x', 'o'], ['b', 'y', 'p'], ['c', 'z', 'q']], columns=['F', 'H', 'G'])
df['G'].replace('p', 'NEW', inplace=True)
```

### Expected Behavior

The replacement being done, and without any warning.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 66e3805b8cabe977f40c05259cc3fcf7ead5687d
python           : 3.10.0.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.11.0-43-generic
Version          : #47~20.04.2-Ubuntu SMP Mon Dec 13 11:06:56 UTC 2021
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.3.5
numpy            : 1.21.2
pytz             : 2021.1
dateutil         : 2.8.2
pip              : 21.3.1
setuptools       : 44.1.1
Cython           : 0.29.24
pytest           : 6.2.5
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.7.1
html5lib         : None
pymysql          : 0.9.3
psycopg2         : 2.9.1 (dt dec pq3 ext lo64)
jinja2           : 3.0.3
IPython          : 7.28.0
pandas_datareader: None
bs4              : 4.10.0
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.3.3
numexpr          : None
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : None
pyarrow          : 6.0.1
pyxlsb           : None
s3fs             : None
scipy            : 1.7.2
sqlalchemy       : 1.2.17
tables           : None
tabulate         : None
xarray           : None
xlrd             : 2.0.1
xlwt             : 1.3.0
numba            : 0.55.0rc1

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #45316
- [x] closes #41062
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

I think this is more intuitive (and "more" correct) than the previous behavior.

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -3481,11 +3481,12 @@ def __getitem__(self, key):
         if is_hashable(key) and not is_iterator(key):
             # is_iterator to exclude generator e.g. test_getitem_listlike
             # shortcut if the key is in columns
-            if self.columns.is_unique and key in self.columns:
-                if isinstance(self.columns, MultiIndex):
-                    return self._getitem_multilevel(key)
+            is_mi = isinstance(self.columns, MultiIndex)
+            if not is_mi and key in self.columns.drop_duplicates(keep=False):
```
Comment (jbrockmendel) [inline pandas/core/frame.py:3485]: comment here pointing back to the relevant issue?

**Comment by jbrockmendel** [comment]
perf impact?

File: pandas/core/frame.py (hunk 1 above)
Comment (phofl) [reply pandas/core/frame.py]: Added

**Comment by phofl** [comment]
> perf impact?

Good point. Let's check for uniqueness or that the key is unique, so we only have the drop duplicates call if we have non unique columns

File: pandas/tests/indexing/test_chaining_and_caching.py [hunk 2]
```diff
@@ -421,19 +421,6 @@ def test_detect_chained_assignment_warnings_errors(self):
             with pytest.raises(com.SettingWithCopyError, match=msg):
                 df.loc[0]["A"] = 111
 
-    def test_detect_chained_assignment_warnings_filter_and_dupe_cols(self):
```
Comment (jbrockmendel) [inline pandas/tests/indexing/test_chaining_and_caching.py:424]: is this test redundant or just the expected behavior changed?

Comment (phofl) [reply pandas/tests/indexing/test_chaining_and_caching.py:424]: Behavior changed, this should be more intuitive now

File: pandas/core/frame.py [hunk 3]
```diff
@@ -3482,11 +3482,17 @@ def __getitem__(self, key):
         if is_hashable(key) and not is_iterator(key):
             # is_iterator to exclude generator e.g. test_getitem_listlike
             # shortcut if the key is in columns
-            if self.columns.is_unique and key in self.columns:
-                if isinstance(self.columns, MultiIndex):
-                    return self._getitem_multilevel(key)
+            is_mi = isinstance(self.columns, MultiIndex)
+            # GH#45316 Return view if key is not duplicated
+            if not is_mi and (
+                self.columns.is_unique
```
Comment (jreback) [inline pandas/core/frame.py:3488]: umm why can't this be just
```not is_mi and key in self.columns``` ?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
@phofl can you merge master and some comments

File: pandas/core/frame.py (hunk 3 above)
Comment (phofl) [reply pandas/core/frame.py:3488]: Because this works only if it returns a series. If self.columns contains key twice, this fails.

**Comment by phofl** [comment]
Done

**Comment by phofl** [comment]
Merged master again

**Comment by mroeschke** [comment]
@phofl could you merge in one more time. Looks pretty good.

**Comment by phofl** [comment]
Done

File: pandas/core/frame.py (hunk 3 above)
Comment (mroeschke) [reply pandas/core/frame.py:3618]: Shouldn't `key in self.columns.drop_duplicates(keep=False)` cover the `is_unique and key in self.columns` case?

Comment (phofl) [reply pandas/core/frame.py:3618]: I am pretty sure I did that for performance reasons

Comment (phofl) [reply pandas/core/frame.py:3619]: Yep already did. Took me way to long to remember that for it being obvious :)

**Review by mroeschke — APPROVED** [review]
LGTM. Looks to need @jreback 's approval to merge

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @phofl very nice as always

**Pull request merged**